File: silt-ts/src/primitives.ts
import { Node, eng, link, propagate } from './engine';

export class Signal<T> extends Node {
    val: T;

    constructor(val: T) {
        super();
        this.val = val;
    }

    get(): T {
        link(this);
        return this.val;
    }

    set(val: T) {
        if (this.val !== val) {
            this.val = val;
            for (const o of this.obs) {
                o.state = 2;
                if (!eng.q.includes(o)) eng.q.push(o);
            }
            propagate();
        }
    }
}

export class Computed<T> extends Node {
    fn: () => T;
    val!: T;
    dirty = true;

    constructor(fn: () => T) {
        super();
        this.fn = fn;
    }

    update() {
        this.state = 0;
        const old = this.val;
        
        for (const s of this.src) s.obs.delete(this);
        this.src.clear();

        const prev = eng.active;
        eng.active = this;
        try {
            this.val = this.fn();
        } finally {
            eng.active = prev;
        }

        if (old !== this.val || this.dirty) {
            this.dirty = false;
            for (const o of this.obs) {
                o.state = 2;
                if (!eng.q.includes(o)) eng.q.push(o);
            }
        }
    }

    get(): T {
        if (this.dirty || this.state > 0) this.update();
        link(this);
        return this.val;
    }
}

export class Effect extends Node {
    fn: () => void;

    constructor(fn: () => void) {
        super();
        this.fn = fn;
        this.update();
    }

    update() {
        this.state = 0;
        for (const s of this.src) s.obs.delete(this);
        this.src.clear();

        const prev = eng.active;
        eng.active = this;
        try {
            this.fn();
        } finally {
            eng.active = prev;
        }
    }
}

File: silt-python/silt/primitives.py
from .engine import Node, eng, link, propagate, emit

class Signal(Node):
    def __init__(self, val):
        super().__init__()
        self.val = val

    def get(self):
        link(self)
        return self.val

    def set(self, val):
        if self.val != val:
            self.val = val
            emit('set', self.id, val)
            # weakset iteration needs list copy
            for o in list(self.obs):
                o.state = 2
                if o not in eng.q:
                    eng.q.append(o)
            propagate()

class Computed(Node):
    def __init__(self, fn):
        super().__init__()
        self.fn = fn
        self.val = None
        self.dirty = True
        
    def update(self):
        self.state = 0
        old = self.val
        for s in self.src:
            s.obs.discard(self)
        self.src.clear()
        
        prev = eng.active
        eng.active = self
        try:
            new_val = self.fn()
        finally:
            eng.active = prev
            
        if old != new_val or self.dirty:
            self.val = new_val
            self.dirty = False
            for o in list(self.obs):
                o.state = 2
                if o not in eng.q:
                    eng.q.append(o)

    def get(self):
        if self.dirty or self.state > 0:
            self.update()
        link(self)
        return self.val

class Effect(Node):
    def __init__(self, fn):
        super().__init__()
        self.fn = fn
        eng.roots.add(self)
        self.update()

    def update(self):
        self.state = 0
        for s in self.src:
            s.obs.discard(self)
        self.src.clear()
        
        prev = eng.active
        eng.active = self
        try:
            self.fn()
        finally:
            eng.active = prev
            
    def stop(self):
        eng.roots.discard(self)
        for s in self.src:
            s.obs.discard(self)
        self.src.clear()